src/Parsers/ExpressionElementParsers.cpp

```diff
@@ -834,10 +826,12 @@ bool ParserNumber::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 
     auto try_read_float = [&](const char * it, const char * end)
     {
+        std::string buf(it, end); /// Copying is needed to ensure the string is 0-terminated.
         char * str_end;
         errno = 0;    /// Functions strto* don't clear errno.
-        Float64 float_value = std::strtod(it, &str_end);
-        if (str_end == end && errno != ERANGE)
+        /// The usage of strtod is needed, because we parse hex floating point literals as well.
```

Comment:
Would it be possible to use https://en.cppreference.com/w/cpp/utility/from_chars instead (and maybe in parseNumber too)?

Reply:
If it is parsing hex float literals 0x9AB1p10 - yes.

Reply:
It seems to work if you pass the appropriate std::chars_format::hex (as we do with hex):

```
$ clickhouse local
ClickHouse local version 22.13.1.1.

Mordor :) SELECT 0x1.234p01;

SELECT 2.275390625

Query id: ed3f3f3e-94c3-4477-82a9-336c4c7e3271

┌─2.275390625─┐
│ 2.275390625 │
└─────────────┘

1 row in set. Elapsed: 0.001 sec. 
```


Removed the initial 0x and passed `std::chars_format::hex`:
```c++
#include <cassert>
#include <charconv>
#include <iomanip>
#include <iostream>
#include <optional>
#include <string_view>
#include <system_error>
 
int main()
{
    for (std::string_view const str : {"1.234p01"}) {
 
        std::cout << "String: " << std::quoted(str) << ". ";
 
        double result{};
 
        auto [ptr, ec] { std::from_chars(str.data(), str.data() + str.size(), result, std::chars_format::hex) };
 
        if (ec == std::errc())
        {
            std::cout << "Result: " << result << '\n';
        }
        else if (ec == std::errc::invalid_argument)
        {
            std::cout << "That isn't a double.\n";
        }
        else if (ec == std::errc::result_out_of_range)
        {
            std::cout << "This number is larger than an double.\n";
        }
    }
}
```

```
/tmp $ clang++ -std=c++20 a.cpp 
/tmp $ ./a.out 
String: "1.234p01". Result: 2.27539
```


No need to copy data and it [should be much faster](https://lemire.me/blog/2022/07/27/comparing-strtod-with-from_chars-gcc-12/), so if it covers all use cases it'd be great for both integers and floats.